core/dataset.cpp

```diff
@@ -1108,4 +1108,79 @@ Dataset histogram(const Dataset &dataset, const Dim &dim) {
   return histogram(dataset, bins);
 }
 
+template <class A, class B>
+void merge_validate_proxies(const A &lhs, const B &rhs) {
+  for (const auto & [ k, v ] : lhs) {
+    /* Throw if an item is present in both proxies but they are not equal */
+    if (rhs.contains(k) && rhs[k] != v) {
+      throw std::runtime_error("Both datasets contain an item with the same "
+                               "key that does not match.");
+    }
+  }
+}
+
+template <class T> void merge_copy_data(Dataset &dest, const T &src) {
+  for (const auto & [ name, data ] : src) {
+    if (!dest.contains(name)) {
+      dest.setData(std::string(name), data);
+    }
+  }
+}
+
+template <class T> void merge_copy_coords(Dataset &dest, const T &src) {
+  for (const auto & [ dim, coord ] : src.coords()) {
+    if (!dest.coords().contains(dim)) {
+      dest.setCoord(dim, coord);
+    }
+  }
+}
+
+template <class T> void merge_copy_labels(Dataset &dest, const T &src) {
+  for (const auto & [ name, label ] : src.labels()) {
+    if (!dest.labels().contains(name)) {
+      dest.setLabels(std::string(name), label);
+    }
+  }
+}
```

Comment:
I think `merge_validate_proxies` (except when used for data) and the various `merge_copy_coords` (etc.) are mostly doing the same as implemented by `union_`, see, e.g., https://github.com/scipp/scipp/blob/1517f0ebc115c25bb5e5e45bc0bd2d05e3a28db2/core/data_array.cpp#L31. I think you can use that here?

Reply:
Yes, good point. I as unaware we had the `union_` function.